<?xml version="1.0" ?>

<robot name="charlie" xmlns:xacro="http://ros.org/wiki/xacro">

  <xacro:include filename="$(find charlie)/urdf/charlie.gazebo.xacro"/>

  <!-- Base link -->
  <link name="base_footprint"/>

  <joint name="base_joint" type="fixed">
    <parent link="base_footprint"/>
    <child link="base_link" />
    <origin xyz="-0.18 0 0.095" rpy="0 0 0"/>
  </joint>
  
  <link name="base_link">
    <visual>
      <geometry>
        <box size="0.44 0.4 0.01"/> <!-- Flat rectangular base -->
      </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <material name="gray">
        <color rgba="0.5 0.5 0.5 1.0"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <box size="0.44 0.4 0.01"/>
      </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
    </collision>
    <inertial>
      <mass value="4.12"/>
      <origin xyz="0 0 0.0015" rpy="0 0 0"/>
      <inertia ixx="0.0556" ixy="0" ixz="0" iyy="0.0596" iyz="0" izz="0.362"/>
    </inertial>
  </link>
  
  <joint name="battery_joint" type="fixed">
    <parent link="base_link"/>
    <child link="battery_link" />
    <origin xyz="0 0 0.045" rpy="0 0 0"/>
  </joint>
  
  <link name="battery_link">
    <visual>
      <geometry>
        <box size="0.14 0.26 0.09"/>
      </geometry>
      <origin xyz="0.0325 0.005 0" rpy="0 0 0"/>
      <material name="gray">
        <color rgba="0.5 0.5 0.5 1.0"/>
      </material>
    </visual>
  </link>
  
  <joint name="sensor_joint" type="fixed">
    <parent link="battery_link"/>
    <child link="sensor_link" />
    <origin xyz="0 -0.005 0.046" rpy="0 0 0"/>
  </joint>
  
  <link name="sensor_link">
    <visual>
      <geometry>
        <box size="0.14 0.31 0.002"/>
      </geometry>
      <origin xyz="0.12 0.005 0" rpy="0 0 0"/>
      <material name="white">
        <color rgba="0.2 0.2 0.2 0.9"/>
      </material>
    </visual>
  </link>

  <!-- Wheels -->
  <link name="wheel_left_link">
    <visual>
      <geometry>
        <cylinder radius="0.085" length="0.055"/>
      </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <material name="black">
        <color rgba="0 0 0 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <cylinder radius="0.085" length="0.055"/>
      </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
    </collision>
    
    <inertial>
      <mass value="0.258"/>
      <inertia ixx="0.000939" ixy="0" ixz="0" iyy="0.000939" iyz="0" izz="0.000939"/>
      <origin xyz="0 0 0" rpy="0 0 0"/>
    </inertial>

  </link>

  <link name="wheel_right_link">
    <visual>
      <geometry>
        <cylinder radius="0.085" length="0.055"/>
      </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <material name="black">
        <color rgba="0 0 0 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <cylinder radius="0.085" length="0.055"/>
      </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
    </collision>
    
    <inertial>
      <mass value="0.258"/>
      <inertia ixx="0.000939" ixy="0" ixz="0" iyy="0.000939" iyz="0" izz="0.000939"/>
      <origin xyz="0 0 0" rpy="0 0 0"/>
    </inertial>
    
  </link>

  <!-- Casters -->
  <link name="left_caster">
    <visual>
      <geometry>
        <sphere radius="0.04"/>
      </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <material name="black">
        <color rgba="0 0 0 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <sphere radius="0.04"/>
      </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
    </collision>
  </link>

  <link name="right_caster">
    <visual>
      <geometry>
        <sphere radius="0.04"/>
      </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <material name="black">
        <color rgba="0 0 0 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <sphere radius="0.04"/>
      </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
    </collision>
  </link>

  <!-- LiDAR -->
  <link name="base_scan">
    <visual>
      <geometry>
        <mesh filename="package://charlie/meshes/sensors/omo_r1_lidar.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <material name="blue">
        <color rgba="0 0 0 1"/>
      </material>
    </visual>
  </link>

  <!-- Joints -->
  <joint name="wheel_left_joint" type="continuous">
    <parent link="base_link"/>
    <child link="wheel_left_link"/>
    <origin xyz="0.18 0.235 -0.01" rpy="-1.57 0 0"/>
    <axis xyz="0 0 1"/>
  </joint>

  <joint name="wheel_right_joint" type="continuous">
    <parent link="base_link"/>
    <child link="wheel_right_link"/>
    <origin xyz="0.18 -0.235 -0.01" rpy="1.57 0 0"/>
    <axis xyz="0 0 -1"/>
  </joint>

  <joint name="left_caster_joint" type="fixed">
    <parent link="base_link"/>
    <child link="left_caster"/>
    <origin xyz="-0.17 0.17 -0.055" rpy="1.57 0 0"/>
  </joint>

  <joint name="right_caster_joint" type="fixed">
    <parent link="base_link"/>
    <child link="right_caster"/>
    <origin xyz="-0.17 -0.17 -0.055" rpy="1.57 0 0"/>
  </joint>

  <joint name="scan_joint" type="fixed">
    <parent link="base_link"/>
    <child link="base_scan"/>
    <origin xyz="0.17 0 0.11" rpy="0 0 0"/>
  </joint>
</robot>

